cdap-app-fabric/src/main/java/io/cdap/cdap/internal/metadata/MetadataConsumerSubscriberService.java

```diff
@@ -0,0 +1,305 @@
+                                            Constants.Metadata.METADATA_WRITER_SUBSCRIBER, messageId);
+  }
+
+  @Override
+  protected boolean shouldRunInSeparateTx(ImmutablePair<String, MetadataMessage> message) {
+    // if this message caused a conflict last time we tried, stop here to commit all messages processed so far
+    if (message.getFirst().equals(conflictMessageId)) {
+      return true;
+    }
+    // operations at the instance or namespace level can take time. Stop here to process in a new transaction
+    EntityType entityType = message.getSecond().getEntityId().getEntityType();
+    return entityType.equals(EntityType.INSTANCE) || entityType.equals(EntityType.NAMESPACE);
+  }
+
+  @Override
+  protected void preProcess() {
+    // no-op
+  }
+
+  @Override
+  protected void doStartUp() throws Exception {
+    super.doStartUp();
+  }
+
+  @Override
+  protected void processMessages(StructuredTableContext structuredTableContext,
+                                 Iterator<ImmutablePair<String, MetadataMessage>> messages)
+    throws IOException, ConflictException {
+    Map<MetadataMessage.Type, MetadataMessageProcessor> processors = new HashMap<>();
+    // Loop over all fetched messages and process them with corresponding MetadataMessageProcessor
+    while (messages.hasNext()) {
+      ImmutablePair<String, MetadataMessage> next = messages.next();
+      String messageId = next.getFirst();
+      MetadataMessage message = next.getSecond();
+      MetadataMessageProcessor processor = processors.computeIfAbsent(message.getType(), type -> {
+        if (type == MetadataMessage.Type.FIELD_LINEAGE) {
+          return new FieldLineageProcessor(this.cConf, this.provider.loadMetadataConsumers());
+        }
+        return null;
+      });
+      // Intellij would warn here that the condition is always false - because the switch above covers all cases.
+      // But if there is ever an unexpected message, we can't throw exception, that would leave the message there.
+      if (processor == null) {
+        LOG.warn("Unsupported metadata message type {}. Message ignored.", message.getType());
+        continue;
+      }
+      try {
+        processor.processMessage(message, structuredTableContext);
+        conflictCount = 0;
+      } catch (ConflictException e) {
+        if (messageId.equals(conflictMessageId)) {
+          conflictCount++;
+          if (conflictCount >= maxRetriesOnConflict) {
+            LOG.warn("Skipping metadata message {} after processing it has caused {} consecutive conflicts: {}",
+                     message, conflictCount, e.getMessage());
+            continue;
+          }
+        } else {
+          conflictMessageId = messageId;
+          conflictCount = 1;
+        }
+        throw e;
+      }
+    }
+  }
+
+  /**
+   * The {@link MetadataMessageProcessor} for processing field lineage.
+   */
+  private static final class FieldLineageProcessor implements MetadataMessageProcessor {
```

Comment:
Can this be an independent class?

Reply:
Kept it inside the main class as it is specific to this implementation